sdk/core/core-http/src/fetchHttpClient.ts

```diff
@@ -195,9 +196,33 @@ export abstract class FetchHttpClient implements HttpClient {
 
       throw fetchError;
     } finally {
-      // clean up event listener
-      if (httpRequest.abortSignal && abortListener) {
-        httpRequest.abortSignal.removeEventListener("abort", abortListener);
+      // clean up event listener. for streaming operations clean up on close/end/error event.
+      const cleanUpStreamListener = (stream: NodeJS.ReadableStream) => {
+        const cleanupCallback = () => {
+          stream.removeListener("close", cleanupCallback);
+          stream.removeListener("end", cleanupCallback);
+          stream.removeListener("error", cleanupCallback);
+          if (abortListener) {
+            httpRequest.abortSignal?.removeEventListener("abort", abortListener);
+          }
+        };
+        stream.once("close", cleanupCallback);
+        stream.once("end", cleanupCallback);
+        stream.once("error", cleanupCallback);
+      };
+
+      // it's super rare if not impossible that both download streaming and upload streaming happen
```

Comment:
This seems a little implausible, but given that anything in NodeJS can come in as a stream (even a very simple request body JSON payload), I don't think it's impossible. Is there a reason we can't handle both?

Reply:
By this rare case I meant, a HTTP request is uploading a stream body while at the same time, streaming down the response body.  I chatted with @bterlson and he doesn't think we need to worry about this case.

Reply:
Oh I think I shouldn't have if-else here.  Both uploading stream and downloading stream should have the clean up call back hooked up.

Reply:
But I need to make sure the last to finish does the clean up.

Reply:
I remember my intuition was that if for a particular request, if there's a download stream, it would probably ends after the upload stream (if there's any), so clean up at the end of downloading, otherwise clean up at the end of uploading stream if any. Both failing, cleaning up when response returns.

Reply:
It's not technically impossible for a response stream to start before a request stream finishes from a conceptual standpoint, but everywhere up and down the stack needs to work in that case and, ultimately, boil up to an application that is comfortable writing a response before a request has even finished. This all seems very rare, and googling seems to confirm.

That said, if it's not hard to support, might not be a bad idea to support it? I imagine a bug here woul dbe very hard to track down...

Reply:
Because we use the same abort signal for upload stream and download stream, we want to keep the listener around until both stream are done.  There are some possible fixes I think:

1. use separate abort signals for upload and download, clean up each one independently, or
2. if we could somehow figured which stream ends last, do the clean up on that one's end/error event.

Reply:
Since we don't have to worry about multi-thread race condition, would simple checks like in https://github.com/Azure/azure-sdk-for-js/pull/12038/commits/c2a8a8a066fd9dbd771a4441e67c4f5cb9500177 work? I am not sure how to test this edge case with real request/response though.

Reply:
I think the current version should work. @xirzec please have another look.